### PR TITLE
yosys: Update to 0.30 and fix TCL detection

### PR DESCRIPTION
**Pull request by hmartinez82**
@umarcor Can you please take a look too 👀

### CODE REVIEW

**Comment by mmuetzel** [comment]
`MINGW32` failed with errors like the following:
```
  D:/M/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/12.2.0/../../../../i686-w64-mingw32/bin/ld.exe: kernel/driver.o: in function `main':
  C:\_\B\src\yosys-yosys-0.26/kernel/driver.cc:574: undefined reference to `_imp__Tcl_EvalFile'
  D:/M/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/12.2.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\_\B\src\yosys-yosys-0.26/kernel/driver.cc:575: undefined reference to `_imp__Tcl_GetStringResult'
```
Maybe an issue with calling conventions that don't match?

**Comment by mmuetzel** [comment]
`UCRT64` failed with the following error:
https://github.com/msys2/MINGW-packages/actions/runs/4289118302/jobs/7471844561#step:11:2019
```
  D:/M/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find D:/a/msys64/ucrt64/lib/gcc/x86_64-w64-mingw32/12.2.0/adalib/libgnat.a: No such file or directory
  collect2.exe: error: ld returned 1 exit status
  make: *** [passes/techmap/Makefile.inc:58: yosys-filterlib.exe] Error 1
```
That library should already be installed at `/ucrt64/lib/gcc/x86_64-w64-mingw32/12.2.0/adalib/libgnat.a`:
https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-gcc-ada?repo=ucrt64
Is it broken for some reason?

**Comment by hmartinez82** [comment]
@mmuetzel I'm going to try to debug the MINGW32 one. And indeed. I'm seeing that hardcoded wrong path in my local UCRT64 builds too :(

**Comment by umarcor** [comment]
@hmartinez82 see #13821. I will bump it. Please submit other modifications as separated PRs, since they are not related to updating the package.

**Comment by hmartinez82** [comment]
@umarcor I did see that PR, but it was missing other parts in the PKGBUILD file, like installing the `COPYING` file. Do you mean these changes like that?

**Comment by hmartinez82** [comment]
@umarcor Awesome. I'll close this PR and wait until you update the version to 0.26 in https://github.com/msys2/MINGW-packages/pull/13821

I'm also curious about these other dependencies in  https://archlinux.org/packages/community/x86_64/yosys/ . Are they just "nice to have". Should I add them in my next PR?

**Comment by umarcor** [comment]
@hmartinez82 #13821 is updated and CI is running now.

With regard to other dependencies, I don't think graphviz or xdot should be installed by default. If users want to use those features, I believe they can install those packages without rebuilding yosys.

With regard to solvers (cvc4, yices and z3) I think those are runtime dependencies as well, for users doing formal verification (through sby). So, it seems that Arch package is building both yosys and symbiyosys at the same time.

**Comment by mmuetzel** [comment]
> Installing `COPYING` is an enhancement to the recipe, unrelated to the issues with updating the version of the tools.

You could still do that in the same PR (maybe not same commit) to avoid needing to bump the pkgrel just for the addition of the license file.

**Comment by hmartinez82** [comment]
@mmuetzel. I found the issue for UCRT64 https://github.com/msys2/MINGW-packages/issues/16034

**Comment by hmartinez82** [comment]
@mmuetzel You were right, the symbols are a mess:
objdump -t /mingw32/lib/libtcl.dll.a | grep Tcl_EvalFile returns:
``` 
[  6](sec  1)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x00000000 _Tcl_EvalFile
[  7](sec  3)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x00000000 __imp__Tcl_EvalFile
```
$ objdump.exe -t  /mingw64/lib/libtcl.dll.a | grep Tcl_EvalFile returns:
```
[  5](sec  1)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 Tcl_EvalFile
[  6](sec  3)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp_Tcl_EvalFile
```
Is that extra underscore the standard behavior? 
And yosys in MINGW32 is expecting:
```
D:\Dev\Github\MINGW-packages\mingw-w64-yosys\src\yosys-yosys-0.26/kernel/driver.cc:574: undefined reference to '_imp__Tcl_EvalFile'
```

**Comment by mmuetzel** [comment]
Those additional underscores are there to mark which calling convention is used when calling those functions across library boarders:
https://learn.microsoft.com/en-us/cpp/build/reference/decorated-names?view=msvc-170

If the calling convention used in the importing library doesn't match the calling convention expected by the exporting library, you might be seeing those `undefined reference to` errors.
I don't know if this is the case here.

**Comment by mmuetzel** [comment]
Those additional underscores are there to mark which calling convention is used when calling those functions across library boarders:
https://learn.microsoft.com/en-us/cpp/build/reference/decorated-names?view=msvc-170

If the calling convention used in the importing library doesn't match the calling convention expected by the exporting library, you might be seeing those `undefined reference to` errors.
I don't know if this is the case here.

**Comment by hmartinez82** [comment]
Ah, I knew about MSVC's name mangling 😊, I just wasn't sure about GCC's.

**Comment by hmartinez82** [comment]
MING32 is probably failing due to https://github.com/msys2/MINGW-packages/issues/16814

**Comment by hmartinez82** [comment]
Added workaround for detecting TCL's version and compiler flags

File: mingw-w64-yosys/001-fix-tcl-flags.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+-ifeq ($(CONFIG),mxe)
+-CXXFLAGS += -DYOSYS_ENABLE_TCL
+-LDLIBS += -ltcl86 -lwsock32 -lws2_32 -lnetapi32 -lz -luserenv
+-else
+-CXXFLAGS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --silence-errors --cflags tcl || echo -I$(TCL_INCLUDE)) -DYOSYS_ENABLE_TCL
+-LDLIBS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --silence-errors --libs tcl || echo $(TCL_LIBS))
+-endif
++TCL_INCLUDE ?= $(PREFIX)/include/$(TCL_VERSION)
++TCL_LIBS ?= -L$(PREFIX)/lib -ltcl
++CXXFLAGS += $(echo -I$(TCL_INCLUDE)) -DYOSYS_ENABLE_TCL
```
Comment (mmuetzel) [inline mingw-w64-yosys/001-fix-tcl-flags.patch:29]: Is there a reason to shell out and call `echo`. Wouldn't the following be the same but avoid spawning a new process?
``` suggestion
+CXXFLAGS += -I$(TCL_INCLUDE) -DYOSYS_ENABLE_TCL
```

Comment (hmartinez82) [reply mingw-w64-yosys/001-fix-tcl-flags.patch:29]: @mmuetzel doh, you are right. I even did that for LDLIBS already!

Comment (hmartinez82) [reply mingw-w64-yosys/001-fix-tcl-flags.patch]: Fixed :)

**Comment by hmartinez82** [comment]
@Biswa96 I'll give it a try after The Merge

File: mingw-w64-yosys/PKGBUILD [hunk 2]
```diff
@@ -1,51 +1,57 @@
...
 arch=('any')
 mingw_arch=('mingw32' 'mingw64' 'ucrt64')
 license=('spdx:ISC')
 url="http://www.clifford.at/yosys/"
 groups=("${MINGW_PACKAGE_PREFIX}-eda")
-depends=(
-  "${MINGW_PACKAGE_PREFIX}-ghdl"
-  "${MINGW_PACKAGE_PREFIX}-python"
-)
+depends=("${MINGW_PACKAGE_PREFIX}-ghdl"
```
Comment (umarcor) [inline mingw-w64-yosys/PKGBUILD:13]: Please, preserve the format, to avoid modifying the lines that are not changed.

File: mingw-w64-yosys/PKGBUILD [hunk 3]
```diff
@@ -1,51 +1,57 @@
+
 _realname=yosys
 pkgbase=mingw-w64-${_realname}
 pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
-pkgver=0.12.r45
+pkgver=0.27
```
Comment (umarcor) [inline mingw-w64-yosys/PKGBUILD:5]: Yosys 0.28 was tagged five days ago: https://github.com/YosysHQ/yosys/releases/tag/yosys-0.28

Comment (hmartinez82) [reply mingw-w64-yosys/PKGBUILD]: Updated to 0.30 :)

File: mingw-w64-yosys/PKGBUILD [hunk 4]
```diff
@@ -12,52 +13,63 @@ groups=("${MINGW_PACKAGE_PREFIX}-eda")
...
-  "${_realname}::git+https://github.com/YosysHQ/${_realname}.git#commit=${_commit}"
-  "ghdl-yosys-plugin::git+https://github.com/ghdl/ghdl-yosys-plugin.git#commit=09a32cd1"
-)
-sha256sums=(
-  'SKIP'
-  'SKIP'
-)
+_commit='f7a8284'
+_plugin_commit='5b64ccf'
+source=("https://github.com/YosysHQ/yosys/archive/refs/tags/${_realname}-${pkgver}.tar.gz"
```
Comment (umarcor) [inline mingw-w64-yosys/PKGBUILD:29]: Why did you change it to use a tarball when downloading yosys but still use git to retrieve ghdl-yosys-plugin?

File: mingw-w64-yosys/001-fix-tcl-flags.patch [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+-endif
+-
+-ifeq ($(CONFIG),mxe)
+-CXXFLAGS += -DYOSYS_ENABLE_TCL
+-LDLIBS += -ltcl86 -lwsock32 -lws2_32 -lnetapi32 -lz -luserenv
+-else
+-CXXFLAGS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --silence-errors --cflags tcl || echo -I$(TCL_INCLUDE)) -DYOSYS_ENABLE_TCL
+-LDLIBS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --silence-errors --libs tcl || echo $(TCL_LIBS))
+-endif
++TCL_INCLUDE ?= $(PREFIX)/include/$(TCL_VERSION)
```
Comment (umarcor) [inline mingw-w64-yosys/001-fix-tcl-flags.patch:27]: Adding the `$(PREFIX)` to https://github.com/YosysHQ/yosys/blob/master/Makefile#L469 seems a sensible contribution to be done upstream.

With regard to TCL_LIBS, it seems that we can set it in the PKGBUILD, without requiring a patch.

File: mingw-w64-yosys/PKGBUILD (hunk 4 above)
Comment (hmartinez82) [reply mingw-w64-yosys/PKGBUILD:29]: Because there's a preference for tarball. Best would be if the plugins repo also provided tags/tarballs. But I can move everything back to git

Comment (umarcor) [reply mingw-w64-yosys/PKGBUILD:29]: It's the same. GitHub provides zipfiles/tarballs for any "ref". For instance: `https://github.com/ghdl/ghdl-yosys-plugin/archive/5b64ccfdeee6c75f70487c1ea153ec3e1fb26cd1.zip`. It works with a shorter hash as well: `https://github.com/ghdl/ghdl-yosys-plugin/archive/5b64ccfd.zip`.

File: mingw-w64-yosys/PKGBUILD [hunk 6]
```diff
@@ -12,52 +13,63 @@ groups=("${MINGW_PACKAGE_PREFIX}-eda")
...
   fi
 
-  if [ "$CARCH" = "x86_64" ]; then
+  if [ "$MSYSTEM_CARCH" = "x86_64" ]; then
     make config-msys2-64
   fi
 
-  mv "${srcdir}/ghdl-yosys-plugin"/src frontends/ghdl
+  mv "${srcdir}/ghdl-yosys-plugin/src" frontends/ghdl
+  local _tclver=$(echo 'set a "[info tclversion]" ; puts $a ; exit' | $MINGW_PREFIX/bin/tclsh)
```
[the remark this answers is not in the document]
Comment (hmartinez82) [reply mingw-w64-yosys/PKGBUILD:54]: Even if it's for clarity?

File: mingw-w64-yosys/001-fix-tcl-flags.patch (hunk 5 above)
Comment (hmartinez82) [reply mingw-w64-yosys/001-fix-tcl-flags.patch:27]: The CLANG32 issue was fixed we could use your original PR if we prefer.

**Comment by hmartinez82** [comment]
I noticed that yosys has already been updated to 0.30. Closing this.
